packages/codec/lib/index.ts

```diff
@@ -16,7 +16,147 @@ require("source-map-support/register");
+   *   1. `type`: This is a [[Format.Types.Type|`Type`]] object describing the value's
+   *     type.  Each `Type` has a `typeClass` field describing the overall broad type,
+   *     such as `"uint"` or `"bytes"`, together with additional information that gives
+   *     the specific type.  For full detail, see [[Format.Types]].
+   *
+   *   2. `kind`: This is either `"value"`, in which case the `Result` is a
+   *     [[Format.Values.Value|`Value`]], or `"error"`, in which case the `Result` is an
+   *     [[Format.Errors.ErrorResult|`ErrorResult`]].  In the former case, there will be
+   *     a `value` field containin the decoded value.  In the latter case, there will be
+   *     an `error` field indicating what went wrong.  *Warning*: When decoding a
+   *     complex type, such as an array, mapping, or array, getting a kind of `"value"`
+   *     does not necessarily mean the individual elements were decoded successfully.
+   *     Even if the `Result` for the array (mapping, struct) as a whole has kind
+   *     `"value"`, the elements might still have kind `"error"`.
+   *
+   *   3. `value`: As mentioned, this is included when `kind` is equal to `"value"`.
+   *     It contains information about the actual decoded value.  See
+   *     [[Format.Values|`Format.Values`]] for more information.
+   *
+   *   4. `error`: The alternative to `value`.  Generally includes information about
+   *     the raw data that led to the error.  See [[Format.Errors|`Format.Errors`]] for
+   *     more information.
+   *
+   *   5. `reference`: This field is a debugger-only feature and does not
+   *      apply to results returned by  @truffle/decoder, so it won't be documented here.
+   *
+   * ### Values vs. errors
+   *
+   * It's worth taking a moment here to answer the question: What counts as a value,
+   * and what counts as an error?
+   *
+   * In general, the answer is that anything that can be generated via Solidity
+   * alone (i.e. no assembly), with correctly-encoded inputs, and without making use
+   * of compiler bugs, is a value, not an error.  That means that, for instance, the
+   * following things are values, not errors:
+   *   - A variable of contract type whose address does not actually hold a
+   *     contract of that type;
+   *   - An external function pointer that does not correspond to a valid
+   *     function;
+   *   - A string containing invalid UTF-8;
+   *   - ..., etc.
+   *
+   * By contrast, the following *are* errors:
+   *   - A `bool` which is neither `false` (0) nor `true` (1);
+   *   - An `enum` which is out of range;
+   *   - ..., etc.
+   *
+   * (You may be wondering about the enum case here, because if you go sufficiently
+   * far back, to Solidity 0.4.4 or earlier, it *was* possible to generate
+   * out-of-range enums without resorting to assembly or compiler bugs.  However,
+   * enums are only supported in full mode, which only supports 0.4.9 and later, so
```

Comment:
The mention of full mode here doesn't make sense anymore now that you've moved this outside the README.  Not sure how to fix this.

Reply:
Saw that too, I'll figure it out

Reply:
Fastest solution here is to move the notes on decoding modes to a particular symbol, likely DecodingMode... if you are not extremely opposed to this, then that can save figuring out how to link to the docs index safely.

(Leaving the breadcrumb of header + content link, of course :)

Thoughts?

Reply:
Hm, OK.  Maybe group the two outsourced sections under a "key concepts" or "important things to be aware" of section with then like two subheadings each with its link? Or something?

Reply:
Mind if I add this task to #2505 and leave for a separate PR?

Reply:
Sounds good to me!